File: src/main/java/bad/robot/radiate/monitor/Observer.java
package bad.robot.radiate.monitor;

import bad.robot.radiate.Activity;
import bad.robot.radiate.Progress;
import bad.robot.radiate.Status;

@Deprecated
public interface Observer {
    default void update(Observable source, Status status) { /* ignore status updates */ }
    default void update(Observable source, Activity activity, Progress progress) { /* ignore status updates */ }
    default void update(Observable source, Information information) { /* ignore status updates */ }
    default void update(Observable source, Exception exception) { /* ignore status updates */ }
}

File: src/main/scala/bad/robot/radiate/monitor/Observer.scala
package bad.robot.radiate.monitor

import bad.robot.radiate._
import org.apache.log4j.Logger
import bad.robot.radiate.Error

trait Observer {
  def update(source: Observable, status: Status)
  def update(source: Observable, activity: Activity, progress: Progress)
  def update(source: Observable, information: Information)
  def update(source: Observable, exception: Exception)
  def update(source: Observable, error: Error)
}

class LoggingObserver extends Observer {
  override def update(source: Observable, status: Status): Unit = () /** ignore status updates **/

  override def update(source: Observable, activity: Activity, progress: Progress): Unit = () /** ignore progress updates **/

  override def update(source: Observable, information: Information) {
    Logger.getLogger(source.getClass).info(information)
  }

  override def update(source: Observable, exception: Exception) {
    Logger.getLogger(source.getClass).error(exception.getMessage, exception)
  }

  override def update(source: Observable, error: Error) {
    Logger.getLogger(source.getClass).error(error.message)
  }
}
